
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(PNWColors)
outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/sebastes_genome/analysis/rer_converge/plot_RER_converge_analyses_final/output/lm_rer"

t_Lmat50 = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/RERconverge/20210318/_Lmat50.named_combined_correlations_rerconverge.20210318.txt",header=T,sep="\t") %>%
  mutate(type="Lmat50")

t_max_depth = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/RERconverge/20210318/_max_depth.named_combined_correlations_rerconverge.20210318.txt",header=T,sep="\t") %>%
  mutate(type="max_depth")

t_resid = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/RERconverge/20210318/_m_depth_matL_residual.named_combined_correlations_rerconverge.20210318.txt",header=T,sep="\t") %>%
  mutate(type="resid")

t_full_lifespan = read.table("/Volumes/GoogleDrive/Shared drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/RERconverge/20210316/lifespan.named_combined_correlations_rerconverge.20210316.txt",header=T,sep="\t") %>%
   mutate(type="all")

```
```{r}
annot_paths = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/Human_genes_categories/curated.c2.all.v7.1.symbols.txt",header=T,sep="\t")

GO_annots = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/Sebastes/manuscript/supplement/individual_analyses/Human_genes_categories/GO.c5.all.v7.1.symbols.txt",header=T,sep="\t",quote="")

all_annots = rbind(annot_paths,GO_annots)
annots = all_annots  %>% 
  rowwise() %>% 
  mutate(genesets = list(strsplit(genesets,"\\|")[[1]]))

```

```{r}

t = rbind(t_Lmat50,t_max_depth,t_resid,t_full_lifespan)
t_sum = t %>% mutate(direction=ifelse(Rho>0,"+","-")) %>%
  group_by(type) %>%
  arrange(P) %>%
  mutate(exp_P=row_number()/n(),
         rnum = row_number()) %>%
  mutate(type = factor(type))

t_D = t_sum %>% 
  dplyr::select(Human_gene,p.adj,type) %>%
  ungroup() %>%
  filter(p.adj<0.05) %>%
  filter(type=="max_depth") %>%
  dplyr::select(Human_gene)

t_R = t_sum %>% 
  dplyr::select(Human_gene,p.adj,type) %>%
  ungroup() %>%
  filter(p.adj<0.05) %>%
  filter(type=="resid") %>%
  dplyr::select(Human_gene)

t_S = t_sum %>% 
  dplyr::select(Human_gene,p.adj,type) %>%
  ungroup() %>%
  filter(p.adj<0.05) %>%
  filter(type=="Lmat50") %>%
  dplyr::select(Human_gene)

t_A = t_sum %>% 
  dplyr::select(Human_gene,p.adj,type) %>%
  ungroup() %>%
  filter(p.adj<0.05) %>%
  filter(type=="all") %>%
  dplyr::select(Human_gene)

write.table(t_A,paste(outdir,"All.txt",sep=""),row.names=FALSE,quote=FALSE)
write.table(t_D,paste(outdir,"Depth.txt",sep=""),row.names=FALSE,quote=FALSE)
write.table(t_R,paste(outdir,"Resid.txt",sep=""),row.names=FALSE,quote=FALSE)
write.table(t_S,paste(outdir,"Size.txt",sep=""),row.names=FALSE,quote=FALSE)


```

```{r}
library(ggrepel)

g_R = c("AKT1", 
        "FOXO1", 
        "IGF1", 
        "ADRB3",
        "TMEM101", #3 
        "TEDC2",
        "RUFY2",
        "LCMT1",
        "GCGR", #2 #glucagon receptor that is important in controlling blood glucose levels
        "CALCR", #1 #calcium homeostasis 
        "GSE1", #oncogene
        "MLXIPL", #glucose dependent TF
        "DYRK2",##DYRK2 -> phosphorylates TERT
        #"SARDH", #enzyme localized to the mitochondrial matrix which catalyzes the oxidative demethylation of sarcosine
        "CAV1", #  initiating step in coupling integrins to the Ras-ERK pathway and promoting cell cycle progression. The gene is a tumor suppressor gene candidate. Negatively regulates TGFB1-mediated activation of SMAD2/3 by mediating the internalization of TGFBR1 from membrane rafts leading to its subsequent degradation 
        "TRAF3IP1", # inhibitor of the innate type I IFN response
        #"NCAPD2", #Regulatory subunit of the condensin complex, a complex required for conversion of interphase chromatin into mitotic-like condense chromosomes. The condensin complex probably introduces positive supercoils into relaxed DNA in the presence of type I topoisomerases and converts nicked DNA into positive knotted forms in the presence of type II topoisomerases
        #"DNJAB2", # protein folding and protein complex dissociation
        #DOCK4 #cytokinesis ??
        "NRG1" #Induces the phosphorylation and activation of MAPK3/ERK1, MAPK1/ERK2 and AKT1.  critical role in the growth and development of multiple organ systems
        #"PAIP2", #translational repression
        #"SLC9A4" #Involved in pH regulation to eliminate acids generated by active metabolism or to counter adverse environmental conditions
) 

#g_S = c("NEK6","NUDT5","BRIP1","VFW","PUM2","MLEC")
#g_D = c("PNLIP","AQP1","LACC1")
#g_O = c("TRAF3IP3","NECTIN3","TP53I3","CYP8B1","RPL23","MADD")

#goi = c(g_R,g_S,g_D,g_O)
goi = c(g_R)

t_sum %>% 
  filter(p.adj<0.05) %>%
  group_by(type) %>%
  summarize(n=n())

mx_y = 30
mn_y = 3
mx_x = 4
mn_x = 0.5

t_sum_label = t_sum %>% 
  filter(type=="resid") %>%
  arrange(P) %>%
  mutate(label = paste(Human_gene," (",row_number(),")",sep="")) %>%
  filter(Human_gene %in% goi) %>%
  mutate(rnum = row_number()) %>%
  mutate(curr_max_logP = max(-log(P,10))) %>%
  ungroup() %>%
  mutate(n_sig = sum(p.adj<0.05),
         delta_y = (mx_y-mn_y)/n_sig,
         delta_x = (mx_x-mn_x)/n_sig,
         rnum_pos_neg = row_number(),
         ypos = (n_sig-rnum_pos_neg)*delta_y+mn_y,
         xpos = (n_sig-rnum_pos_neg)*delta_x+mn_x)

t_sum$type = factor(t_sum$type, levels = c("all","Lmat50", "max_depth", "resid"))

colors = pnw_palette("Sunset2",5)[c(1,3,5)]
qq_colors = c("grey",colors)

g=ggplot(t_sum %>% arrange(type))
g=g+
  theme_bw()+
  geom_abline(a=0,b=1,alpha=.8)+
  scale_x_continuous(expression(-log[10](expected)))+
  scale_y_continuous(expression(-log[10](observed)))+
  geom_point(aes(x=-log(exp_P,10),y=-log(P,10),color=type),fill='white',alpha=.7,size=0.2,shape=16)+ #,shape=p.adj<0.05
  geom_point(aes(x=-log(exp_P,10),y=-log(P,10),color=type),
             shape=21,
             size=.25,
             alpha=1,
             data = t_sum %>% arrange(type) %>% filter(p.adj<0.05))+
  geom_label(aes(x=xpos,
                y=ypos,
                label=label),
             size=2,
             color='black',
             fill=alpha(qq_colors[4],.8),
             label.padding = unit(0.1, "lines"),
             label.size=NA,
             data = t_sum_label,
             show.legend = FALSE,
            hjust=1)+
  geom_segment(aes(x=xpos,xend=-log(exp_P,10),
                   y=ypos,yend=-log(P,10),
                color=type),
            show.legend = FALSE,
            size=.2,
            alpha=1,
            data=t_sum_label)+
  theme_bw(base_size=10)+
  scale_color_manual("",values=qq_colors,labels=c("lifespan (91)","Size at Maturity (66)", "Maximum Depth (49)", "Model Residual (56)"))+
  scale_fill_manual("",values=qq_colors,labels=c("lifespan (91)","Size at Maturity (66)", "Maximum Depth (49)", "Model Residual (56)"))+
  theme(legend.position=c(0.25,0.92),
        legend.key.size=unit(.15,"cm"),
        legend.background = element_blank(),
        axis.title = element_text(size=8))+
  guides(color=guide_legend(override.aes = list(size=2)))

pdf(paste(outdir,"/RER_qq_genes.pdf",sep=""),width=3,height=2.5)
print(g)
dev.off()

png(paste(outdir,"/RER_qq_genes.png",sep=""),width=3,height=2.5,units="in",res=600)
print(g)
dev.off()

```
```{r}

set_all = (t_sum %>% filter(p.adj<0.05) %>% filter(type=="all"))$Human_gene
set_D = (t_sum %>% filter(p.adj<0.05) %>% filter(type=="max_depth"))$Human_gene
set_R = (t_sum %>% filter(p.adj<0.05) %>% filter(type=="resid"))$Human_gene
set_L = (t_sum %>% filter(p.adj<0.05) %>% filter(type=="Lmat50"))$Human_gene

length(intersect(set_all,set_D))
length(intersect(set_all,set_L))
length(intersect(set_all,set_R))

#TOTAL lifespan associated genes
length(union(set_all,set_R))

```
```{r}

mx_y = 30
mn_y = 0.5

mx_x = 3
mn_x = 0

t_sum_dir = t %>% mutate(direction=ifelse(Rho>0,"+","-")) %>%
  group_by(type,direction) %>%
  arrange(P) %>%
  mutate(exp_P=row_number()/n(),
         rnum = row_number()) %>%
  mutate(type = factor(type)) %>%
  ungroup() %>%
  group_by(type) %>%
  mutate(curr_max_logP = max(-log(P,10))) %>%
  ungroup() %>%
  group_by(type,direction) %>%
  mutate(n_sig = sum(p.adj<0.05),
         delta_y = (curr_max_logP-mn_y)/n_sig,
         delta_x = (mx_x-mn_x)/n_sig,
         rnum_pos_neg = row_number(),
         ypos = (n_sig-rnum_pos_neg)*delta_y+mn_y,
         xpos = (n_sig-rnum_pos_neg)*delta_x+mn_x-(ifelse(direction=="+",.5,0)))
  
g2=ggplot(t_sum_dir %>% arrange(type))
g2=g2+
  theme_bw()+
  geom_abline(a=0,b=1,alpha=.8)+
  scale_x_continuous(expression(-log[10](expected)))+
  scale_y_continuous(expression(-log[10](observed)))+
  geom_point(aes(x=-log(exp_P,10),
                 y=-log(P,10),
                 color=direction),alpha=.7,size=0.2)+ 
  geom_segment(aes(x=xpos,xend=-log(exp_P,10),
                   y=ypos,yend=-log(P,10),
                color=direction),
            show.legend = FALSE,
            size=.1,
            alpha=.2,
            data=t_sum_dir %>%
                 filter(p.adj<0.05))+
  geom_text(aes(x=xpos,
                y=ypos,
                color=direction,
                label=Human_gene),
            show.legend = FALSE,
            hjust=0,
            vjust=0,
            size=1,
            data=t_sum_dir %>%
                 filter(p.adj<0.05))+
  theme_bw(base_size=10)+
  scale_color_manual("",values=colors)+
  theme(legend.position=c(0.05,0.95),
        legend.key.size=unit(.15,"cm"),
        legend.background = element_blank(),
        axis.title = element_text(size=8))+
  guides(color=guide_legend(override.aes = list(size=2)))+
  facet_wrap(~type,scales="free")+
  theme(strip.background = element_blank())

png(paste(outdir,"/RER_qq_genes_pos_minus.png",sep=""),width=4,height=4,units="in",res=600)
print(g2)
dev.off()

```

```{r}
library(colorspace)

get_ternary_color = function(data, colors, V1, V2, V3){
  C1 = hex2RGB(colors[1])
  C2 = hex2RGB(colors[2])
  C3 = hex2RGB(colors[3])
  data %>% 
    mutate(R = floor((C1@coords[1]*{{V1}}+C2@coords[1]*{{V2}}+C3@coords[1]*{{V3}})*255),
           G = floor((C1@coords[2]*{{V1}}+C2@coords[2]*{{V2}}+C3@coords[2]*{{V3}})*255), 
           B = floor((C1@coords[3]*{{V1}}+C2@coords[3]*{{V2}}+C3@coords[3]*{{V3}})*255)) %>%
    mutate(color=rgb(R,G,B,maxColorValue=255)) %>%
    dplyr::select(-c(R,G,B))
}

```

```{r}
library(ggtern)
library(tricolore)

t_sum_full= t_full_lifespan %>%
  mutate(full_P=P, full_adj_P = p.adj) %>% 
  dplyr::select(Genes,full_P,full_adj_P) 

t_sum_full %>% filter(full_adj_P<0.05)

t_tern = t_sum %>% 
  filter(type != "all") %>%
  mutate(logP = -log(P)) %>%
  group_by(Genes,Human_gene) %>%
  mutate(frac = logP/sum(logP),
         best_P = min(P),
         best_adjP = min(p.adj)) %>%
  dplyr::select(Genes,frac,type,best_P,best_adjP) %>%
  pivot_wider(names_from = type, values_from=frac) %>%
  inner_join(t_sum_full,by="Genes") #%>%
t_tern = get_ternary_color(t_tern, colors, Lmat50,max_depth, resid)

pdf(paste(outdir,"/ternary_RER_genes_contributions_to_all.pdf",sep=""),width=4,height=2.15)

ggtern(data=t_tern %>% filter((full_adj_P<0.05)))+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color),alpha=1,shape=21)+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color),alpha=.6)+
  geom_Tline(Tintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_Rline(Rintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_Lline(Lintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_mask()+
  theme_bw(base_size=6)+
  theme_showarrows()+
  theme(legend.key.size=unit(0.25,"cm"))+
  scale_color_identity()+
  labs(Tarrow="residual",
       Rarrow="maximum depth",
       Larrow="size at maturity")+
  scale_L_continuous("(S)ize",breaks=c(.25,.50,.75,1))+
  scale_R_continuous("(D)epth",breaks=c(.25,.50,.75,1))+
  scale_T_continuous("(R)esidual",breaks=c(.25,.50,.75,1))+
  scale_size_continuous(expression(-log[10](P)),range=c(2,5))
dev.off()

pdf(paste(outdir,"/ternary_RER_genes_individual_contributions.pdf",sep=""),width=4,height=2.15)
ggtern(data=t_tern %>% filter((best_adjP<0.05)))+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color),alpha=1,shape=21)+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color),alpha=.6)+
  geom_mask()+
  theme_bw(base_size=6)+
  theme_showarrows()+
  theme(legend.key.size=unit(0.25,"cm"))+
  #theme_clockwise()+
  scale_color_identity()+
  labs(Tarrow="residual",
       Rarrow="maximum depth",
       Larrow="size at maturity")+
  scale_L_continuous("(S)ize")+
  scale_R_continuous("(D)epth")+
  scale_T_continuous("(R)esidual")+
  scale_size_continuous(expression(-log[10](P)),range=c(2,5))
dev.off()
```

```{r}
g_R = c("GCN1","NRG1","DNAJB2","ARFGAP1","ADRB3")
g_S = c("NEK6","NUDT5","BRIP1","VFW","PUM2","MLEC")
g_D = c("PNLIP","AQP1","LACC1")
g_O = c("TRAF3IP3","NECTIN3","TP53I3","CYP8B1","RPL23","MADD","AKT1")

goi = c(g_R,g_S,g_D,g_O)

pdf(paste(outdir,"/ternary_RER_genes_contributions_to_all_highlight.pdf",sep=""),width=3,height=2.15)

df = t_tern %>% filter((full_adj_P<0.05)) %>% mutate(goi=Human_gene %in% goi)
ggtern(data=df)+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color,alpha=goi),shape=21)+
  geom_point(aes(Lmat50,resid,max_depth,size=-log(full_P,10),color=color,alpha=goi))+
  geom_text(aes(Lmat50,resid,max_depth,label=Human_gene),
            alpha=1,
            size=1,
            shape=21, 
            data=df %>% filter(goi==TRUE))+
  geom_Tline(Tintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_Rline(Rintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_Lline(Lintercept=0.5,color='black',linetype='dashed',alpha=.8)+
  geom_mask()+
  theme_bw(base_size=12)+
  theme(legend.key.size=unit(0.25,"cm"))+
  scale_color_identity()+
  scale_alpha_manual(values=c(.2,.8))+
  scale_L_continuous("S",breaks=c(),minor_breaks = c())+
  scale_R_continuous("D",breaks=c(),minor_breaks = c())+
  scale_T_continuous("R",breaks=c(),minor_breaks = c())+
  scale_size_continuous(expression(-log[10](P)),range=c(2,5))
dev.off()

```
```{r}
library("viridis")
plot_tern_path = function(outdir,t_tern_data, set_name, annots, title){

  fn = set_name
  set_names=c(set_name)
  currset = annots %>% 
    unnest(genesets) %>%
    filter(geneset.names %in% set_names) %>%
    dplyr::select(genesets)
  
    t_genes_in_set = t_tern_data %>% filter(Human_gene %in% list(currset$genesets)[[1]])
  
    l = dim(t_genes_in_set)[1]
    name = paste(title," (",l,")",sep="")
    
    Lmax = 0.9
    Lmin = (1-Lmax)/2
    br=5
    sbreaks = seq(Lmin,Lmax,(Lmax-Lmin)/br)
    slab = seq(0,100,(100)/br)
    slab = rep("",br+1)
    g=ggtern(data=t_genes_in_set 
                %>% mutate(S=Lmat50+.1,
                           D=max_depth+.1,
                           R=resid+.1))+
      stat_density_tern(geom='polygon',
                             n = 200,
                             bdl = 0,
                             aes(S,R,D,
                                 weight=-log(full_P,10),
                                 fill = ..level..,
                                 alpha = ..level..
                                 )) +
      geom_mask()+
      theme_bw(base_size=4)+
      theme(legend.key.size=unit(0.25,"cm"))+
      scale_fill_viridis(option="magma",direction=1)+
      scale_color_viridis(option="magma",direction=1)+
      scale_L_continuous("S",minor_breaks=c(),breaks=sbreaks,labels=slab,limits = c(Lmin,Lmax))+
      scale_R_continuous("D",minor_breaks=c(),breaks=sbreaks,labels=slab,limits = c(Lmin,Lmax))+
      scale_T_continuous("R",minor_breaks=c(),breaks=sbreaks,labels=slab,limits = c(Lmin,Lmax))+
      theme(legend.position="None",
            plot.title = element_text(hjust=0.5,size=rel(1)))+
    ggtitle(name)+
    guides(alpha=FALSE)

  pdf(paste(outdir,"/pathways_tern/final/ternary_RER_pathway_",fn,".pdf",sep=""),width=1,height=1)
    print(g)
  dev.off()
}

t_tern_data = t_tern

plot_tern_path(outdir,t_tern_data, "KEGG_MTOR_SIGNALING_PATHWAY", annots, "mTOR Signalling")
plot_tern_path(outdir,t_tern_data, "REACTOME_TELOMERE_MAINTENANCE", annots, "Telomere Maintenance")
plot_tern_path(outdir,t_tern_data, "KEGG_ACUTE_MYELOID_LEUKEMIA", annots, "Acute Myeloid Leukemia")
plot_tern_path(outdir,t_tern_data, "REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA", annots, "Response to Hypoxia")
plot_tern_path(outdir,t_tern_data, "KEGG_INSULIN_SIGNALING_PATHWAY", annots, "Insulin Signalling")
plot_tern_path(outdir,t_tern_data, "KEGG_RIBOSOME", annots, "Ribosome")
plot_tern_path(outdir,t_tern_data, "GO_HEAT_SHOCK_PROTEIN_BINDING", annots, "Heat Shock Protein Binding")
plot_tern_path(outdir,t_tern_data, "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION", annots, "Antigen Processing & Presentation")
plot_tern_path(outdir,t_tern_data, "KEGG_GLYCEROLIPID_METABOLISM", annots, "Glycerolipid metabolism")
plot_tern_path(outdir,t_tern_data, "KEGG_APOPTOSIS", annots, "Apoptosis")
plot_tern_path(outdir,t_tern_data, "KEGG_FATTY_ACID_METABOLISM", annots, "Fatty Acid Metabolism")
plot_tern_path(outdir,t_tern_data, "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION", annots, "Vasopressin regulated\nwater reabsorption")



```

